[linux-6.6.y] efi: cper: Add Zhaoxin/Centaur ZDI/ZPI error decode

leoliu-oc opened the pull request:
zhaoxin inclusion
category: feature

-------------------

ZPI is the interconnection interface between sockets, ZDI is the interconnection interface between dies.

ZPI is the interconnection interface between sockets, ZDI is the interconnection interface between dies.

When either zdi or zpi occurs error, it will trigger smi interrput, the smi handler will read error information from the zdi/zpi configuration space, fill it in the cper structure asscoiated with error and produce a sci or nmi interrput to notify the OS ,the OS driver will decode the cper structure to help user to annalyze the error.

Because UEFI spec does not define the section type of ZDI/ZPI error. Zhaoxin defines ZDI/ZPI errors according to the error format defined by the Generic Processor Error Section type.When the error occurs, The BIOS will fill error information  in the data structure corresponding to the Generic Processor Error Section type in the smi handler.However,the error information printed by default  apei driver is not easy to read.

The software has added some printed logs to make the ZDI/ZPI error information on the Zhaoxin/Centaur cpu vendor easier to read.

opsiff commented:
在arm64平台有下面的编译错误 需要解决

drivers/firmware/efi/cper.c: In function ‘cper_print_proc_generic’:
drivers/firmware/efi/cper.c:235:13: error: ‘boot_cpu_data’ undeclared (first use in this function); did you mean ‘boot_cpu_init’?
235 |         if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN ||
      |             ^~~~~~~~~~~~~
      |             boot_cpu_init
 drivers/firmware/efi/cper.c:235:13: note: each undeclared identifier is reported only once for each function it appears in
   CC [M]  drivers/net/wireless/realtek/rtl8xxxu/rtl8xxxu_8710b.o
 drivers/firmware/efi/cper.c:235:41: error: ‘X86_VENDOR_ZHAOXIN’ undeclared (first use in this function); did you mean ‘PCI_VENDOR_ID_ZHAOXIN’?
   235 |         if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN ||
       |                                         ^~~~~~~~~~~~~~~~~~
       |                                         PCI_VENDOR_ID_ZHAOXIN
 drivers/firmware/efi/cper.c:236:41: error: ‘X86_VENDOR_CENTAUR’ undeclared (first use in this function); did you mean ‘X86_VENDOR_ANY’?
   236 |             boot_cpu_data.x86_vendor == X86_VENDOR_CENTAUR)
       |                                         ^~~~~~~~~~~~~~~~~~
       |                                         X86_VENDOR_ANY
   CC      drivers/clocksource/arm_arch_timer.o
 make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1
 make[4]: *** [scripts/Makefile.build:480：drivers/firmware/efi] 错误 2
 make[3]: *** [scripts/Makefile.build:480：drivers/firmware] 错误 2

leoliu-oc committed:
efi: cper: Add Zhaoxin/Centaur ZDI/ZPI error decode

zhaoxin inclusion
category: feature

-------------------

ZPI is the interconnection interface between sockets, ZDI is the
interconnection interface between dies.

ZPI is the interconnection interface between sockets, ZDI is the
interconnection interface between dies.

When either zdi or zpi occurs error, it will trigger smi interrput, the
smi handler will read error information from the zdi/zpi configuration
space, fill it in the cper structure asscoiated with error and produce a
sci or nmi interrput to notify the OS ,the OS driver will decode the cper
structure to help user to annalyze the error.

Because UEFI spec does not define the section type of ZDI/ZPI error.
Zhaoxin defines ZDI/ZPI errors according to the error format defined by
the Generic Processor Error Section type.When the error occurs, The BIOS
will fill error information  in the data structure corresponding to the
Generic Processor Error Section type in the smi handler.However,the error
information printed by default  apei driver is not easy to read.

The software has added some printed logs to make the ZDI/ZPI error
information on the Zhaoxin/Centaur cpu vendor easier to read.

Signed-off-by: leoliu-oc <leoliu-oc@zhaoxin.com>

leoliu-oc commented:
> 在arm64平台有下面的编译错误 需要解决
> 
> drivers/firmware/efi/cper.c: In function ‘cper_print_proc_generic’: drivers/firmware/efi/cper.c:235:13: error: ‘boot_cpu_data’ undeclared (first use in this function); did you mean ‘boot_cpu_init’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~ | boot_cpu_init drivers/firmware/efi/cper.c:235:13: note: each undeclared identifier is reported only once for each function it appears in CC [M] drivers/net/wireless/realtek/rtl8xxxu/rtl8xxxu_8710b.o drivers/firmware/efi/cper.c:235:41: error: ‘X86_VENDOR_ZHAOXIN’ undeclared (first use in this function); did you mean ‘PCI_VENDOR_ID_ZHAOXIN’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~~~~~~ | PCI_VENDOR_ID_ZHAOXIN drivers/firmware/efi/cper.c:236:41: error: ‘X86_VENDOR_CENTAUR’ undeclared (first use in this function); did you mean ‘X86_VENDOR_ANY’? 236 | boot_cpu_data.x86_vendor == X86_VENDOR_CENTAUR) | ^~~~~~~~~~~~~~~~~~ | X86_VENDOR_ANY CC drivers/clocksource/arm_arch_timer.o make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1 make[4]: *** [scripts/Makefile.build:480：drivers/firmware/efi] 错误 2 make[3]: *** [scripts/Makefile.build:480：drivers/firmware] 错误 2

感谢指正。新版中已经处理了这个问题。

opsiff commented:
> > 在arm64平台有下面的编译错误 需要解决
> > drivers/firmware/efi/cper.c: In function ‘cper_print_proc_generic’: drivers/firmware/efi/cper.c:235:13: error: ‘boot_cpu_data’ undeclared (first use in this function); did you mean ‘boot_cpu_init’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~ | boot_cpu_init drivers/firmware/efi/cper.c:235:13: note: each undeclared identifier is reported only once for each function it appears in CC [M] drivers/net/wireless/realtek/rtl8xxxu/rtl8xxxu_8710b.o drivers/firmware/efi/cper.c:235:41: error: ‘X86_VENDOR_ZHAOXIN’ undeclared (first use in this function); did you mean ‘PCI_VENDOR_ID_ZHAOXIN’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~~~~~~ | PCI_VENDOR_ID_ZHAOXIN drivers/firmware/efi/cper.c:236:41: error: ‘X86_VENDOR_CENTAUR’ undeclared (first use in this function); did you mean ‘X86_VENDOR_ANY’? 236 | boot_cpu_data.x86_vendor == X86_VENDOR_CENTAUR) | ^~~~~~~~~~~~~~~~~~ | X86_VENDOR_ANY CC drivers/clocksource/arm_arch_timer.o make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1 make[4]: *** [scripts/Makefile.build:480：drivers/firmware/efi] 错误 2 make[3]: *** [scripts/Makefile.build:480：drivers/firmware] 错误 2
> 
> 感谢指正。新版中已经处理了这个问题。
drivers/firmware/efi/cper.c:173:13: error: ‘cper_print_proc_generic_zdi_zpi’ defined but not used [-Werror=unused-function]
 173 | static void cper_print_proc_generic_zdi_zpi(const char *pfx,
     |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 cc1: all warnings being treated as errors
 make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1

arm64上还有一个unused的问题 改完就应该可以了

leoliu-oc commented:
> > > 在arm64平台有下面的编译错误 需要解决
> > > drivers/firmware/efi/cper.c: In function ‘cper_print_proc_generic’: drivers/firmware/efi/cper.c:235:13: error: ‘boot_cpu_data’ undeclared (first use in this function); did you mean ‘boot_cpu_init’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~ | boot_cpu_init drivers/firmware/efi/cper.c:235:13: note: each undeclared identifier is reported only once for each function it appears in CC [M] drivers/net/wireless/realtek/rtl8xxxu/rtl8xxxu_8710b.o drivers/firmware/efi/cper.c:235:41: error: ‘X86_VENDOR_ZHAOXIN’ undeclared (first use in this function); did you mean ‘PCI_VENDOR_ID_ZHAOXIN’? 235 | if (boot_cpu_data.x86_vendor == X86_VENDOR_ZHAOXIN || | ^~~~~~~~~~~~~~~~~~ | PCI_VENDOR_ID_ZHAOXIN drivers/firmware/efi/cper.c:236:41: error: ‘X86_VENDOR_CENTAUR’ undeclared (first use in this function); did you mean ‘X86_VENDOR_ANY’? 236 | boot_cpu_data.x86_vendor == X86_VENDOR_CENTAUR) | ^~~~~~~~~~~~~~~~~~ | X86_VENDOR_ANY CC drivers/clocksource/arm_arch_timer.o make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1 make[4]: *** [scripts/Makefile.build:480：drivers/firmware/efi] 错误 2 make[3]: *** [scripts/Makefile.build:480：drivers/firmware] 错误 2
> > 
> > 
> > 感谢指正。新版中已经处理了这个问题。
> > drivers/firmware/efi/cper.c:173:13: error: ‘cper_print_proc_generic_zdi_zpi’ defined but not used [-Werror=unused-function]
> > 173 | static void cper_print_proc_generic_zdi_zpi(const char *pfx,
> > |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > cc1: all warnings being treated as errors
> > make[5]: *** [scripts/Makefile.build:243：drivers/firmware/efi/cper.o] 错误 1
> 
> arm64上还有一个unused的问题 改完就应该可以了

该问题已经解决。还烦请再review了，谢谢。
